File: scraper/scrape_idealo.ts
import { chromium } from "playwright-extra";
import StealthPlugin from "puppeteer-extra-plugin-stealth";
import * as cheerio from "cheerio";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

chromium.use(StealthPlugin());

// idealo.de Apple Mac mini / Mac Studio category
const URL =
  "https://www.idealo.de/preisvergleich/ProductCategory/3020F189906-105543532-107083121-107083122-107083123.html?sortKey=minPrice";

interface IdealoProduct {
  name: string;
  model: string;
  price: number;
  priceFormatted: string;
  imageUrl: string;
  productUrl: string;
  offerCount: number;
  cpu: string;
  ram: string;
  storage: string;
  gpu: string;
  os: string;
  formFactor: string;
  description: string;
}

function parsePrice(text: string): { price: number; formatted: string } {
  // idealo prices: "2.589,00 €"
  const match = text.match(/([\d.]+),(\d{2})\s*€/);
  if (!match) return { price: 0, formatted: text };
  const price = Number.parseFloat(
    match[1].replace(/\./g, "") + "." + match[2],
  );
  return { price, formatted: match[0].trim() };
}

function parseOfferCount(text: string): number {
  // "1 Angebot" or "4 Angebote"
  const match = text.match(/(\d+)\s*Angebot/);
  return match ? Number.parseInt(match[1], 10) : 0;
}

function extractSpecs(name: string, text: string): {
  ram: string;
  storage: string;
  cpu: string;
  gpu: string;
  os: string;
  formFactor: string;
  description: string;
} {
  let ram = "";
  let storage = "";
  let cpu = "";
  let gpu = "";
  let os = "";
  let formFactor = "";
  let description = "";

  // Determine form factor from name
  if (/Mac\s*mini/i.test(name)) {
    formFactor = "Mini-PC";
  } else if (/Mac\s*Studio/i.test(name)) {
    formFactor = "Multimedia-PC";
  } else if (/iMac/i.test(name)) {
    formFactor = "All-in-One PC";
  }

  // Extract RAM: "64 GB RAM", "128 GB RAM"
  const ramMatch = text.match(/(\d+)\s*GB\s*RAM/i);
  if (ramMatch) ram = `${ramMatch[1]} GB`;

  // Extract storage: "512 GB SSD", "1.000 GB SSD", etc.
  const storageMatch = text.match(
    /([\d.]+)\s*GB\s*(SSD|Flash)[-]?Speicher/i,
  );
  if (storageMatch) storage = `${storageMatch[1]} GB ${storageMatch[2]}`;

  // Extract CPU: "Apple M4 Pro, 14-Core", "Apple M4 Max, 16-Core", etc.
  const cpuMatch = text.match(
    /Apple\s*(M\d+\s*(Pro|Max|Ultra)?)(,\s*\d+[-]?Core)/i,
  );
  if (cpuMatch) cpu = `Apple ${cpuMatch[1].trim()}${cpuMatch[3] || ""}`;

  // Extract GPU: "Apple M4 Pro 20-Core GPU", "Apple M4 Max 40-Core GPU"
  const gpuMatch = text.match(
    /Apple\s*(M\d+\s*(Pro|Max|Ultra)?)\s*(\d+[-]?Core)\s*GPU/i,
  );
  if (gpuMatch) gpu = `Apple ${gpuMatch[1]} ${gpuMatch[3]} GPU`;

  // Extract OS
  const osMatch = text.match(/macOS\s*(Sequoia|Ventura|Monterey|Sonoma)/i);
  if (osMatch) os = `macOS ${osMatch[1]}`;

  // Description is the remaining descriptive text
  const cleaned = text
    .replace(ramMatch?.[0] || "", "")
    .replace(storageMatch?.[0] || "", "")
    .replace(cpuMatch?.[0] || "", "")
    .replace(gpuMatch?.[0] || "", "")
    .replace(osMatch?.[0] || "", "")
    .replace(/\s+/g, " ")
    .trim();
  description = cleaned;

  return { ram, storage, cpu, gpu, os, formFactor, description };
}

async function scrapeIdealo() {
  console.log("Launching browser with stealth for idealo.de...");
  const browser = await chromium.launch({
    headless: true,
    args: [
      "--no-sandbox",
      "--disable-blink-features=AutomationControlled",
      "--disable-features=IsolateOrigins,site-per-process",
      "--disable-site-isolation-trials",
    ],
  });
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
    viewport: { width: 1920, height: 1080 },
    locale: "de-DE",
    timezoneId: "Europe/Berlin",
  });
  const page = await context.newPage();

  // Additional stealth
  await page.addInitScript(() => {
    Object.defineProperty(navigator, "webdriver", { get: () => false });
    // @ts-ignore
    window.chrome = { runtime: {} };
  });

  console.log("Navigating to idealo.de...");
  await page.goto(URL, { waitUntil: "domcontentloaded", timeout: 30000 });

  // Wait for content to load
  await page.waitForTimeout(4000);

  // Scroll down slowly to trigger lazy loading
  await page.evaluate(async () => {
    await new Promise<void>((resolve) => {
      let totalHeight = 0;
      const distance = 350;
      const timer = setInterval(() => {
        window.scrollBy(0, distance);
        totalHeight += distance;
        if (totalHeight >= document.body.scrollHeight) {
          clearInterval(timer);
          resolve();
        }
      }, 200);
    });
  });

  await page.waitForTimeout(2000);

  // Try to click "accept cookies" if present
  try {
    const cookieBtn = page.locator(
      'button:has-text("Alle akzeptieren"), button:has-text("Accept all"), button:has-text("Zustimmen")',
    );
    if ((await cookieBtn.count()) > 0) {
      await cookieBtn.first().click();
      await page.waitForTimeout(1500);
      console.log("Accepted cookies");
    }
  } catch {
    console.log("No cookie banner found or already accepted");
  }

  const html = await page.content();

  // Debug: check what we got
  const bodyPreview = await page.evaluate(() =>
    document.body.innerText.slice(0, 800),
  );
  console.log("Body preview:", bodyPreview.slice(0, 400));

  await browser.close();

  // Save full HTML for debugging
  Bun.write("debug_idealo.html", html);

  console.log("\nParsing HTML...");
  const $ = cheerio.load(html);

  const products: IdealoProduct[] = [];
  const priceRegex = /([\d.]+),(\d{2})\s*€/;

  // idealo typically uses various container classes
  // Try multiple strategies to find product cards

  // Strategy 1: Look for offer-list items (idealo classic)
  const offerListItems = $(
    '[class*="offerList"], [class*="productOffers"], [class*="srp-offer"], [data-offer-id]',
  ).toArray();
  console.log(`Offer list items: ${offerListItems.length}`);

  // Strategy 2: Look for product cards (idealo newer layout)
  const productCards = $(
    'article, [class*="product"], [class*="Product"], [class*="item"]',
  ).toArray();
  console.log(`Product card candidates: ${productCards.length}`);

  // Strategy 3: Scan ALL elements for price-like patterns and Mac names
  const allElements = $("div, article, li, section, tr").toArray();
  const candidates: { el: cheerio.AnyNode; text: string }[] = [];
  const seenTexts = new Set<string>();

  for (const el of allElements) {
    const text = $(el).text().trim();
    if (
      text.length > 60 &&
      text.length < 3000 &&
      priceRegex.test(text) &&
      /Apple|Mac mini|Mac Studio|iMac/i.test(text) &&
      !seenTexts.has(text.slice(0, 120))
    ) {
      seenTexts.add(text.slice(0, 120));
      candidates.push({ el, text });
    }
  }

  console.log(`Found ${candidates.length} product candidates`);

  // Filter: prefer leaf nodes (most specific)
  const leafCandidates = candidates.filter((c) => {
    const containedByOther = candidates.some(
      (other) =>
        other !== c &&
        $(c.el).parents().toArray().includes(other.el),
    );
    // We want the most specific (not parent of another)
    const isParent = candidates.some(
      (other) =>
        other !== c &&
        $(other.el).parents().toArray().includes(c.el),
    );
    return !isParent;
  });

  console.log(`Filtered to ${leafCandidates.length} leaf candidates`);

  // Deduplicate by name/model
  const uniqueCandidates: typeof leafCandidates = [];
  const seenNames = new Set<string>();

  for (const c of leafCandidates) {
    // Extract a unique key from the name
    const nameMatch = c.text.match(
      /(Apple\s*(Mac\s*(mini|Studio)))\s*(M\d|20\d\d)/i,
    );
    const modelMatch = c.text.match(/\((Z1[A-Z0-9\-_]+)\)/i);
    const key = modelMatch
      ? modelMatch[1]
      : nameMatch
        ? c.text.slice(0, 60).trim()
        : c.text.slice(0, 40);
    if (!seenNames.has(key)) {
      seenNames.add(key);
      uniqueCandidates.push(c);
    }
  }

  console.log(`Deduplicated to ${uniqueCandidates.length} unique products`);

  for (const { el, text } of uniqueCandidates) {
    const $el = $(el);

    // --- Name ---
    // idealo product name pattern
    let name =
      $el.find('[class*="productTitle"], [class*="offerTitle"], [class*="title"], h2, h3')
        .first()
        .text()
        .trim() ||
      text.split("\n")[0]?.trim() ||
      "";

    // If name is too short, try to extract the full product name
    if (name.length < 15) {
      const longName = text.match(/Apple\s*(Mac\s*(mini|Studio))\s*(20\d\d|M\d).{5,100}?(?=\d{1,3}(?:\.\d{3})*,\d{2}\s*€)/i);
      if (longName) name = longName[0].trim();
    }

    // --- Model number ---
    const modelMatch = text.match(/\((Z1[A-Z0-9\-_]+)\)/);
    const model = modelMatch ? modelMatch[1] : "";

    // --- Price ---
    // idealo shows the lowest price
    const allPrices = [...text.matchAll(new RegExp(priceRegex, "g"))];
    let price = 0;
    let priceFormatted = "";
    if (allPrices.length > 0) {
      // First price is usually the main one
      const p = parsePrice(allPrices[0]![0]);
      price = p.price;
      priceFormatted = p.formatted;
    }
    if (price === 0) continue;

    // --- Offer count ---
    const offerMatch = text.match(/(\d+)\s*Angebot/);
    const offerCount = offerMatch ? Number.parseInt(offerMatch[1], 10) : 0;

    // --- Image ---
    const img = $el.find("img").first();
    const imageUrl = img.attr("src") || img.attr("data-src") || img.attr("data-lazy-src") || "";

    // --- Product URL ---
    const link = $el.find("a[href]").first();
    let productUrl = "";
    if (link.attr("href")) {
      const href = link.attr("href")!;
      productUrl = href.startsWith("http")
        ? href
        : "https://www.idealo.de" + href;
    }

    // --- Combine all text for spec extraction ---
    const allText = text + " " + name;

    const specs = extractSpecs(name, allText);

    products.push({
      name: name.replace(/\s+/g, " ").trim(),
      model,
      price,
      priceFormatted,
      imageUrl,
      productUrl,
      offerCount,
      ...specs,
    });
  }

  console.log(`\nExtracted ${products.length} products from idealo.de:\n`);
  for (const p of products) {
    console.log(
      `${p.priceFormatted.padEnd(14)} | ${p.name.slice(0, 60).padEnd(62)} | ${p.cpu.padEnd(25)} | ${p.ram.padEnd(10)} | ${p.storage.padEnd(15)} | ${p.offerCount} offers`,
    );
  }

  return products;
}

scrapeIdealo()
  .then((products) => {
    // Write both to scraper dir and src/data
    const scraperPath = resolve(__dirname, "products_idealo.json");
    Bun.write(scraperPath, JSON.stringify(products, null, 2));
    console.log(`\nWritten ${products.length} products to ${scraperPath}`);

    const srcPath = resolve(__dirname, "..", "src", "data", "products_idealo.json");
    Bun.write(srcPath, JSON.stringify(products, null, 2));
    console.log(`Written ${products.length} products to ${srcPath}`);
  })
  .catch((err) => {
    console.error("idealo scraping failed:", err);
    process.exit(1);
  });

File: scraper/scrape_idealo.py
"""
idealo.de scraper using Camoufox (Firefox-based stealth browser).
Scrapes Apple Mac mini / Mac Studio product listings.
"""
import json
import re
import sys
import time
from pathlib import Path

from camoufox.sync_api import Camoufox
from bs4 import BeautifulSoup

URL = (
    "https://www.idealo.de/preisvergleich/ProductCategory/"
    "3020F189906-105543532-107083121-107083122-107083123.html?sortKey=minPrice"
)

# Output paths
SCRAPER_DIR = Path(__file__).resolve().parent
SRC_DATA_DIR = SCRAPER_DIR.parent / "src" / "data"


def parse_price(text: str) -> tuple[float, str]:
    """Parse idealo price like '2.589,00 €' or 'ab3.238,99 €'"""
    # Match patterns like "2.589,00 €" or "ab3.238,99 €"
    match = re.search(r"((?:ab\s*)?[\d.]+,\d{2}\s*€)", text)
    if not match:
        return 0.0, text.strip()
    raw = match.group(1).strip()
    # Remove "ab" prefix and "€" suffix
    cleaned = raw.replace("ab", "").replace("€", "").strip()
    # Convert German number format
    cleaned = cleaned.replace(".", "").replace(",", ".")
    try:
        price = float(cleaned)
    except ValueError:
        return 0.0, raw
    return price, raw


def parse_offer_count(text: str) -> int:
    """Parse '1 Angebot' or '4 Angebote'"""
    match = re.search(r"(\d+)\s*Angebot", text)
    return int(match.group(1)) if match else 0


def extract_specs(name: str, full_text: str) -> dict:
    """Extract specs from the product text."""
    ram = ""
    storage = ""
    cpu = ""
    gpu = ""
    os = ""
    form_factor = ""

    # Form factor
    if re.search(r"Mac\s*mini", name, re.IGNORECASE):
        form_factor = "Mini-PC"
    elif re.search(r"Mac\s*Studio", name, re.IGNORECASE):
        form_factor = "Multimedia-PC"
    elif re.search(r"iMac", name, re.IGNORECASE):
        form_factor = "All-in-One PC"

    # RAM
    ram_match = re.search(r"(\d+)\s*GB\s*RAM", full_text)
    if ram_match:
        ram = f"{ram_match.group(1)} GB"

    # Storage
    storage_match = re.search(
        r"([\d.]+)\s*GB\s*(SSD|Flash)[-]?Speicher", full_text
    )
    if storage_match:
        storage = f"{storage_match.group(1)} GB {storage_match.group(2)}"

    # CPU
    cpu_match = re.search(
        r"Apple\s*(M\d+\s*(Pro|Max|Ultra)?(?:,\s*\d+[-]?Core)?)", full_text
    )
    if cpu_match:
        cpu = f"Apple {cpu_match.group(1).strip()}"

    # GPU
    gpu_match = re.search(
        r"Apple\s*(M\d+\s*(Pro|Max|Ultra)?)\s*(\d+[-]?Core)\s*GPU", full_text
    )
    if gpu_match:
        gpu = f"Apple {gpu_match.group(1)} {gpu_match.group(3)} GPU"

    # OS
    os_match = re.search(r"macOS\s*(Sequoia|Ventura|Monterey|Sonoma)", full_text)
    if os_match:
        os = f"macOS {os_match.group(1)}"

    return {
        "ram": ram,
        "storage": storage,
        "cpu": cpu,
        "gpu": gpu,
        "os": os,
        "formFactor": form_factor,
    }


def scrape() -> list[dict]:
    products = []

    print("Launching Camoufox browser (Firefox-based stealth)...")
    with Camoufox(
        headless=True,
        os="macos",
        humanize=True,
        geoip=True,
        locale="de-DE",
        screen=None,  # auto-generate
        window=(1920, 1080),
        block_images=False,
        block_webrtc=True,
    ) as browser:
        page = browser.new_page()
        page.set_default_timeout(30000)

        print(f"Navigating to {URL}...")
        page.goto(URL, wait_until="domcontentloaded")

        # Wait for content to render
        time.sleep(3)

        # Accept cookies if present
        try:
            for btn_text in ["Alle akzeptieren", "Accept all", "Zustimmen"]:
                btn = page.query_selector(f'button:has-text("{btn_text}")')
                if btn:
                    btn.click()
                    time.sleep(1.5)
                    print("Accepted cookies")
                    break
        except Exception:
            print("No cookie banner found or already accepted")

        # Scroll to load lazy content
        print("Scrolling to load all products...")
        for _ in range(10):
            page.evaluate("window.scrollBy(0, 400)")
            time.sleep(0.3)

        # Scroll back to top
        page.evaluate("window.scrollTo(0, 0)")
        time.sleep(1)

        # Get page content
        html = page.content()

        # Debug: save HTML
        debug_path = SCRAPER_DIR / "debug_idealo.html"
        debug_path.write_text(html, encoding="utf-8")
        print(f"Saved debug HTML to {debug_path}")

        # Check if we got blocked
        body_text = page.evaluate("document.body.innerText")
        print(f"Body preview: {body_text[:400]}")

        if "Something has gone wrong" in body_text or "Sorry" in body_text[:200]:
            print("ERROR: Blocked by idealo.de bot protection!")
            return products

    print("\nParsing HTML with BeautifulSoup...")
    soup = BeautifulSoup(html, "html.parser")

    price_regex = re.compile(r"([\d.]+),(\d{2})\s*€")

    # Strategy 1: Look for structured offer items
    # idealo uses datalist-datasets with product data
    offer_items = soup.select(
        '[class*="offerList"], [class*="productOffers"], '
        '[class*="srp-offer"], [data-offer-id], '
        '[class*="product-list"] article, '
        '[class*="product-list"] [class*="item"]'
    )
    print(f"Offer list items: {len(offer_items)}")

    # Strategy 2: Scan all divs, articles, li, sections for product patterns
    all_elements = soup.find_all(["div", "article", "li", "section"])
    candidates = []
    seen_texts = set()

    for el in all_elements:
        text = el.get_text(strip=True)
        if (
            60 < len(text) < 3000
            and price_regex.search(text)
            and ("Apple" in text and ("Mac mini" in text or "Mac Studio" in text or "iMac" in text))
            and text[:120] not in seen_texts
        ):
            seen_texts.add(text[:120])
            candidates.append(el)

    print(f"Found {len(candidates)} product candidates")

    # Filter to leaf nodes (most specific containers)
    leaf_candidates = []
    for c in candidates:
        # Check if any other candidate is a descendant of this one
        is_parent = False
        for other in candidates:
            if other is not c and other in c.descendants:
                is_parent = True
                break
        if not is_parent:
            leaf_candidates.append(c)

    print(f"Filtered to {len(leaf_candidates)} leaf candidates")

    # Deduplicate by model name
    unique = []
    seen_models = set()

    for el in leaf_candidates:
        text = el.get_text(strip=True)
        # Try to extract model number as unique key
        model_match = re.search(r"\((Z1[A-Z0-9\-_]+)\)", text)
        if model_match:
            key = model_match.group(1)
        else:
            # Use first 60 chars as key
            key = text[:60].strip()
        if key not in seen_models:
            seen_models.add(key)
            unique.append(el)

    print(f"Deduplicated to {len(unique)} unique products")

    for el in unique:
        text = el.get_text(strip=True)

        # --- Name ---
        name_el = el.find(["h2", "h3", "a"]) or el.find(
            class_=re.compile(r"title|name|product", re.I)
        )
        name = name_el.get_text(strip=True) if name_el else ""

        # If name is still empty or too short, extract from text
        if len(name) < 10:
            name_match = re.search(
                r"Apple\s*(Mac\s*(mini|Studio))\s*(20\d{2}|M\d).*?(?=ab\s*[\d.]+,\d{2}\s*€|[\d.]+,\d{2}\s*€)",
                text,
            )
            if name_match:
                name = name_match.group(0).strip()

        if not name:
            # Fallback: first meaningful line
            lines = [l.strip() for l in text.split("\n") if len(l.strip()) > 10]
            name = lines[0] if lines else text[:80]

        # Clean up name
        name = re.sub(r"\s+", " ", name).strip()

        # --- Model ---
        model_match = re.search(r"\((Z1[A-Z0-9\-_]+)\)", text)
        model = model_match.group(1) if model_match else ""

        # --- Price ---
        all_prices = price_regex.findall(text)
        price = 0.0
        price_formatted = ""
        if all_prices:
            euros, cents = all_prices[0]
            price = float(euros.replace(".", "") + "." + cents)
            price_formatted = f"{all_prices[0][0]},{all_prices[0][1]} €"

        if price == 0.0:
            # Try ab-prefix prices
            price_raw, price_formatted = parse_price(text)
            price = price_raw

        if price == 0.0:
            continue

        # --- Offer count ---
        offer_count = 0
        offer_match = re.search(r"(\d+)\s*Angebot", text)
        if offer_match:
            offer_count = int(offer_match.group(1))

        # --- Image ---
        img = el.find("img")
        image_url = ""
        if img:
            image_url = img.get("src") or img.get("data-src") or img.get("data-lazy-src") or ""

        # --- Product URL ---
        product_url = ""
        link = el.find("a", href=True)
        if link:
            href = link["href"]
            product_url = href if href.startswith("http") else f"https://www.idealo.de{href}"

        # --- Specs ---
        combined_text = f"{name} {text}"
        specs = extract_specs(name, combined_text)

        products.append({
            "name": name,
            "model": model,
            "price": price,
            "priceFormatted": price_formatted,
            "imageUrl": image_url,
            "productUrl": product_url,
            "offerCount": offer_count,
            **specs,
        })

    return products


def main():
    try:
        products = scrape()

        if not products:
            print("\n⚠️  No products scraped. Check debug_idealo.html")
            return

        print(f"\nExtracted {len(products)} products from idealo.de:\n")
        print(f"{'Price':<14} | {'Name':<62} | {'CPU':<28} | {'RAM':<10} | {'Storage':<15} | Offers")
        print("-" * 145)
        for p in products:
            print(
                f"{p['priceFormatted']:<14} | "
                f"{p['name'][:60]:<62} | "
                f"{p['cpu'][:26]:<28} | "
                f"{p['ram']:<10} | "
                f"{p['storage']:<15} | "
                f"{p['offerCount']} offers"
            )

        # Write output files
        scraper_path = SCRAPER_DIR / "products_idealo.json"
        with open(scraper_path, "w", encoding="utf-8") as f:
            json.dump(products, f, indent=2, ensure_ascii=False)
        print(f"\nWritten {len(products)} products to {scraper_path}")

        SRC_DATA_DIR.mkdir(parents=True, exist_ok=True)
        src_path = SRC_DATA_DIR / "products_idealo.json"
        with open(src_path, "w", encoding="utf-8") as f:
            json.dump(products, f, indent=2, ensure_ascii=False)
        print(f"Written {len(products)} products to {src_path}")

    except Exception as e:
        print(f"Scraping failed: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
